File: src/bin/cli/cmd/add.rs
use crate::cli::OutputFormat;
use crate::dto::{AddResultDto, SubtitleDto};
use crate::json_output::{output_error, output_success};
use crate::utils;
use lib::backup::ports::BackupService;
use lib::backup::service::SubRipBackupService;
use lib::subtitle::model::{Subtitle, SubtitleText, SubtitleTimestamp};
use lib::subtitle::ports::SubtitleService;
use lib::subtitle::service::SubRipService;
use log::{debug, error, info};
use serde::Serialize;

#[derive(Serialize)]
struct AddDryRunDto {
    subtitle: SubtitleDto,
    #[serde(flatten)]
    result: AddResultDto,
}

pub fn handle(
    file: &str,
    timestamps: &str,
    text: &str,
    dry_run: bool,
    format: &OutputFormat,
) -> anyhow::Result<()> {
    info!("adding subtitle to file: {}", file);

    // Validate text input
    utils::reject_control_chars(text, "text")?;

    let resolved = utils::resolve_new_path(file)?;

    // Parse timestamps
    debug!("parsing timestamps: {}", timestamps);
    let parts: Vec<&str> = timestamps.splitn(2, '-').collect();

    if parts.len() != 2 {
        output_error(
            format,
            "invalid_timestamp_format",
            "Invalid timestamp format (expected 'HH:MM:SS,mmm-HH:MM:SS,mmm')",
            Some("example: \"00:00:10,000-00:00:12,500\""),
        );
        return Err(anyhow::anyhow!(""));
    }

    let start_str = parts[0];
    let end_str = parts[1];

    let start_duration = Subtitle::parse_timestamp(start_str)
        .map_err(|e| anyhow::anyhow!("invalid start timestamp: {}", e))?;
    let start_timestamp = SubtitleTimestamp::try_new(start_duration)
        .map_err(|e| anyhow::anyhow!("invalid start timestamp value: {}", e))?;

    let end_duration = Subtitle::parse_timestamp(end_str)
        .map_err(|e| anyhow::anyhow!("invalid end timestamp: {}", e))?;
    let end_timestamp = SubtitleTimestamp::try_new(end_duration)
        .map_err(|e| anyhow::anyhow!("invalid end timestamp value: {}", e))?;

    let subtitle_text = SubtitleText::try_new(text.to_string())
        .map_err(|e| anyhow::anyhow!("invalid text: {}", e))?;

    if dry_run {
        info!("dry-run mode, previewing add");
        let service = SubRipService::new(resolved.base_dir);
        let existing = service.get_all(&resolved.filename).unwrap_or_default();
        let next_index = existing
            .iter()
            .map(|s| *s.index.as_ref())
            .max()
            .unwrap_or(0)
            + 1;

        let preview_idx = lib::subtitle::model::SubtitleIndex::try_new(next_index)
            .unwrap_or_else(|_| lib::subtitle::model::SubtitleIndex::try_new(1).unwrap());

        let preview = Subtitle::new(preview_idx, start_timestamp, end_timestamp, subtitle_text)
            .map_err(|e| anyhow::anyhow!("invalid subtitle: {}", e))?;

        let dto = AddDryRunDto {
            subtitle: SubtitleDto::from_subtitle(&preview),
            result: AddResultDto {
                new_index: next_index,
                total_subtitles: existing.len() + 1,
                backup_path: "N/A (dry-run)".into(),
                dry_run: true,
            },
        };

        output_success(format, &dto, || {
            println!("Dry-run: subtitle would be added");
            println!();
            println!("New index: {}", next_index);
            println!("Total subtitles: {}", existing.len() + 1);
            println!();
            println!("{}", preview);
        });

        return Ok(());
    }

    // Create backup
    let backup_service = SubRipBackupService::new();
    let backup_result = backup_service.create_backup(&resolved.full_path);

    let backup_path = match backup_result {
        Ok(Some(path)) => {
            debug!("backup created: {}", path);
            path
        }
        Ok(None) => {
            debug!("file does not exist, skipping backup");
            "N/A (new file)".to_string()
        }
        Err(e) => {
            error!("failed to create backup: {}", e);
            output_error(
                format,
                "backup_failed",
                &format!("Failed to create backup: {}", e),
                None,
            );
            return Err(anyhow::anyhow!(""));
        }
    };

    let service = SubRipService::new(resolved.base_dir);

    debug!("adding new subtitle...");
    match service.add(
        &resolved.filename,
        start_timestamp,
        end_timestamp,
        subtitle_text,
    ) {
        Ok(report) => {
            info!(
                "subtitle added successfully with index {}",
                report.new_index
            );

            let dto = AddResultDto {
                new_index: report.new_index,
                total_subtitles: report.total_subtitles,
                backup_path: backup_path.clone(),
                dry_run: false,
            };

            output_success(format, &dto, || {
                println!("✓ Subtitle added successfully");
                println!();
                println!("New index: {}", report.new_index);
                println!("Total subtitles: {}", report.total_subtitles);
                println!("Backup: {}", backup_path);
            });

            Ok(())
        }
        Err(e) => {
            let cli_err = utils::format_subtitle_error(&e, file);
            output_error(
                format,
                &cli_err.code,
                &cli_err.message,
                cli_err.hint.as_deref(),
            );
            Err(anyhow::anyhow!(""))
        }
    }
}

File: internal/cli/cmd/add.go
package cmd

import (
	"errors"
	"fmt"
	"log/slog"
	"strings"

	"github.com/lebe-dev/submarine/internal/cli"
	"github.com/lebe-dev/submarine/pkg/backup"
	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// addDryRunDto is a port of the Rust `struct AddDryRunDto`. The Rust struct uses
// `#[serde(flatten)]` on the AddResultDto, so its fields are merged at the top
// level alongside the nested `subtitle` object.
type addDryRunDto struct {
	Subtitle cli.SubtitleDto `json:"subtitle"`
	// Flattened AddResultDto fields:
	NewIndex       uint32 `json:"new_index"`
	TotalSubtitles int    `json:"total_subtitles"`
	BackupPath     string `json:"backup_path"`
	DryRun         bool   `json:"dry_run,omitempty"`
}

// HandleAdd is a 1-to-1 port of the Rust `cmd::add::handle`.
func HandleAdd(
	file string,
	timestamps string,
	text string,
	dryRun bool,
	format cli.OutputFormat,
) error {
	slog.Info(fmt.Sprintf("adding subtitle to file: %s", file))

	// Validate text input
	if err := cli.RejectControlChars(text, "text"); err != nil {
		return err
	}

	resolved, err := cli.ResolveNewPath(file)
	if err != nil {
		return err
	}

	// Parse timestamps
	slog.Debug(fmt.Sprintf("parsing timestamps: %s", timestamps))
	parts := strings.SplitN(timestamps, "-", 2)

	if len(parts) != 2 {
		hint := "example: \"00:00:10,000-00:00:12,500\""
		cli.OutputError(
			format,
			"invalid_timestamp_format",
			"Invalid timestamp format (expected 'HH:MM:SS,mmm-HH:MM:SS,mmm')",
			&hint,
		)
		return errors.New("")
	}

	startStr := parts[0]
	endStr := parts[1]

	startDuration, err := subtitle.ParseTimestamp(startStr)
	if err != nil {
		return fmt.Errorf("invalid start timestamp: %s", err)
	}
	startTimestamp, err := subtitle.NewSubtitleTimestamp(startDuration)
	if err != nil {
		return fmt.Errorf("invalid start timestamp value: %s", err)
	}

	endDuration, err := subtitle.ParseTimestamp(endStr)
	if err != nil {
		return fmt.Errorf("invalid end timestamp: %s", err)
	}
	endTimestamp, err := subtitle.NewSubtitleTimestamp(endDuration)
	if err != nil {
		return fmt.Errorf("invalid end timestamp value: %s", err)
	}

	subtitleText, err := subtitle.NewSubtitleText(text)
	if err != nil {
		return fmt.Errorf("invalid text: %s", err)
	}

	if dryRun {
		slog.Info("dry-run mode, previewing add")
		service := subtitle.NewSubRipService(resolved.BaseDir)
		existing, getErr := service.GetAll(resolved.Filename)
		if getErr != nil {
			existing = nil
		}
		var maxIndex uint32
		for _, s := range existing {
			if s.Index.Value() > maxIndex {
				maxIndex = s.Index.Value()
			}
		}
		nextIndex := maxIndex + 1

		previewIdx, err := subtitle.NewSubtitleIndex(nextIndex)
		if err != nil {
			previewIdx, _ = subtitle.NewSubtitleIndex(1)
		}

		preview, err := subtitle.NewSubtitle(previewIdx, startTimestamp, endTimestamp, subtitleText)
		if err != nil {
			return fmt.Errorf("invalid subtitle: %s", err)
		}

		dto := addDryRunDto{
			Subtitle:       cli.SubtitleDtoFromSubtitle(&preview),
			NewIndex:       nextIndex,
			TotalSubtitles: len(existing) + 1,
			BackupPath:     "N/A (dry-run)",
			DryRun:         true,
		}

		cli.OutputSuccess(format, dto, func() {
			fmt.Println("Dry-run: subtitle would be added")
			fmt.Println()
			fmt.Printf("New index: %d\n", nextIndex)
			fmt.Printf("Total subtitles: %d\n", len(existing)+1)
			fmt.Println()
			fmt.Printf("%s\n", preview.String())
		})

		return nil
	}

	// Create backup
	backupService := backup.NewSubRipBackupService()
	backupResult, backupErr := backupService.CreateBackup(resolved.FullPath)

	var backupPath string
	switch {
	case backupErr != nil:
		slog.Error(fmt.Sprintf("failed to create backup: %s", backupErr))
		cli.OutputError(
			format,
			"backup_failed",
			fmt.Sprintf("Failed to create backup: %s", backupErr),
			nil,
		)
		return errors.New("")
	case backupResult == nil:
		slog.Debug("file does not exist, skipping backup")
		backupPath = "N/A (new file)"
	default:
		slog.Debug(fmt.Sprintf("backup created: %s", *backupResult))
		backupPath = *backupResult
	}

	service := subtitle.NewSubRipService(resolved.BaseDir)

	slog.Debug("adding new subtitle...")
	report, err := service.Add(
		resolved.Filename,
		startTimestamp,
		endTimestamp,
		subtitleText,
	)
	if err != nil {
		var subErr *subtitle.SubtitleError
		if errors.As(err, &subErr) {
			cliErr := cli.FormatSubtitleError(subErr, file)
			cli.OutputError(
				format,
				cliErr.Code,
				cliErr.Message,
				cliErr.Hint,
			)
			return errors.New("")
		}
		return err
	}

	slog.Info(fmt.Sprintf("subtitle added successfully with index %d", report.NewIndex))

	dto := cli.AddResultDto{
		NewIndex:       report.NewIndex,
		TotalSubtitles: report.TotalSubtitles,
		BackupPath:     backupPath,
		DryRun:         false,
	}

	cli.OutputSuccess(format, dto, func() {
		fmt.Println("✓ Subtitle added successfully")
		fmt.Println()
		fmt.Printf("New index: %d\n", report.NewIndex)
		fmt.Printf("Total subtitles: %d\n", report.TotalSubtitles)
		fmt.Printf("Backup: %s\n", backupPath)
	})

	return nil
}
